### PR TITLE
Allow managing room knocks

### PR DESCRIPTION
**Pull request by charlynguyen**
We @nordeck are currently implementing the knock rooms behind the feature flag `feature_ask_to_join` (introduced in https://github.com/matrix-org/matrix-react-sdk/pull/11182).

This pull request allows managing room knocks.
- Introduce "People" tab in room settings dialog for knock rooms where all requests to join can be managed 
- Approve requests to join
- Deny requests to join
- Approve and deny require sufficient power levels or the action buttons will be disabled in the UI
- Join request reasons are expandable/collapsable

#### Relates to

- https://github.com/matrix-org/matrix-react-sdk/pull/11353
  - https://github.com/matrix-org/matrix-js-sdk/pull/3647
- https://github.com/matrix-org/matrix-react-sdk/pull/11248
- https://github.com/matrix-org/matrix-react-sdk/pull/11182
- https://github.com/vector-im/element-web/pull/25923

#### Room Settings Tab With Requests to Join

01-people-room-settings-tab

#### Join Request Reasons Are Expandable/Collapsable

02-people-room-settings-tab-see-less

#### Disabled Approve/Deny Buttons

03-people-room-settings-tab-actions-disabled

#### Demo

https://github.com/matrix-org/matrix-react-sdk/assets/1422657/59af3f4c-b9b9-4670-8649-9539ba6114ee

## Checklist

-   [x] Tests written for new code (and old code if feasible)
-   [x] Linter and other CI checks pass
-   [x] Sign-off given on the changes (see [CONTRIBUTING.md](https://github.com/matrix-org/matrix-react-sdk/blob/develop/CONTRIBUTING.md))

---
Here's what your changelog entry will look like:

## ✨ Features
 * Allow managing room knocks ([\#11404](https://github.com/matrix-org/matrix-react-sdk/pull/11404)). Contributed by @charlynguyen.

### CODE REVIEW

**Comment by charlynguyen** [comment]
FYI @daniellekirkwood 🙂

File: src/components/views/settings/tabs/room/PeopleRoomSettingsTab.tsx [hunk 1]
```diff
@@ -0,0 +1,175 @@
...
+            description: error.message,
+        });
+
+    useEffect(() => {
+        const updateKnockMembers = (): void => setKnockMembers(room.getMembersWithMembership("knock"));
+        client.on(RoomStateEvent.Members, updateKnockMembers);
+        return () => {
+            client.removeListener(RoomStateEvent.Members, updateKnockMembers);
+        };
+    }, [client, room]);
```
Comment (t3chguy) [inline src/components/views/settings/tabs/room/PeopleRoomSettingsTab.tsx:151]: Should we use `useTypedEventEmitterState` here instead?

File: src/components/views/dialogs/RoomSettingsDialog.tsx [hunk 2]
```diff
@@ -132,6 +134,16 @@ class RoomSettingsDialog extends React.Component<IProps, IState> {
                 "RoomSettingsGeneral",
             ),
         );
+        if (SettingsStore.getValue("feature_ask_to_join") && this.state.room.getJoinRule() === "knock") {
```
Comment (t3chguy) [inline src/components/views/dialogs/RoomSettingsDialog.tsx:137]: The tab doesn't appear/disappear correctly when the push rule is changed within the lifetime of this dialog being open

**Review by t3chguy — COMMENTED** [review]
Looks generally sane

File: src/components/views/settings/tabs/room/PeopleRoomSettingsTab.tsx (hunk 1 above)
Comment (charlynguyen) [reply src/components/views/settings/tabs/room/PeopleRoomSettingsTab.tsx:151]: Thanks, for pointing that out 🙂

File: src/components/views/dialogs/RoomSettingsDialog.tsx (hunk 2 above)
Comment (charlynguyen) [reply src/components/views/dialogs/RoomSettingsDialog.tsx:137]: Nice catch!

**Review by t3chguy — APPROVED** [review]
Looks sane to me, thanks!

**Comment by daniellekirkwood** [comment]
@charlynguyen catching up on this PR now -- are there no people in the room in the demo video you shared? The designs show the whole people list in the new people tab but they're not showing in your screens :)

**Pull request merged**